### PR TITLE
Gene query tests

### PR DESCRIPTION
**Pull request by wxli0**

### CODE REVIEW

File: src/server/mapping/mapping.js [hunk 1]
```diff
@@ -1,16 +1,18 @@
 
 const fs = require("fs");
 const path = require("path");
-const jsonData = fs.readFileSync(path.resolve(__dirname, "Validation_test_data_set.json"));
+const jsonData = fs.readFileSync(path.resolve(__dirname, "HGNC_dataset.json"));
```
Comment (d2fong) [inline src/server/mapping/mapping.js]: Move this file to gene query folder

File: src/server/gene-query/mapping.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const jsonData = fs.readFileSync(path.resolve(__dirname, "HGNC.json"));
+const jsonContent = JSON.parse(jsonData);
+
+
+// map HGNC symbol to HGNC ID
+// After validation step, we can assume all symbols are valid
+// so that we can find a mapping for sure
+const mapping = function(hgncSymbol) {
+  let id = null;
+  jsonContent.forEach(element => {
```
Comment (d2fong) [inline src/server/gene-query/mapping.js]: It probably makes sense to transform jsonContent into a map https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

so you can just lookup symbols.

File: src/server/gene-query/validate.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  listOfToken.forEach(element => {
+    if(!mapping(element.symbol)) {
+      element.legal = false;
+    } else {
+      element.id = mapping(element);
+    }
+  });
+
+  // check duplicates
+  listOfToken.forEach(element => {
```
Comment (d2fong) [inline src/server/gene-query/validate.js]: delegate collection processing to lodash  https://lodash.com/docs/4.17.4#uniqBy

File: src/server/gene-query/mapping.js (hunk 2 above)
Comment (wxli0) [reply src/server/gene-query/mapping.js]: Since we have many objects, do you mean an array of map?

Comment (d2fong) [reply src/server/gene-query/mapping.js]: right now you are iterating through each jsonContent element and checking if element['approved_symbol'] matches the given hgncSymbol

what you could do is let jsonContent be a map so you have constant time lookup for a given symbol

e.g

```js
const jsonData = JSON.parse(s.readFileSync(path.resolve(__dirname, "HGNC.json")));
const jsonContent = new Map();
jsonContent.forEach(el => jsonContent.set(el['Approved_Symbol'], element));

// then in mapping
const validSymbol = jsonContent.has(hgncSymbol)

If (validSymbol) {
 ...
} else {
...
}

**Pull request merged**